### PR TITLE
C++: Rewrite `cpp/uncontrolled-process-operation` to not use `DefaultTaintTracking`

### PR DESCRIPTION
**Pull request by jketema**
Note that one of the internal tests also needs updating. I'll do that once we're happy with what we have here.

I added a barrier looking at arithmetic types, as these seemed to be the source of many results that are in the end not very interesting. Note that this does mean that we lose some results where the input buffer is copied character-by-character. However, even with the barrier disabled we lose some of these. I'm not sure how worried we should be about this. 

Summary of MRVA results:
* Total number of results reported by MRVA before on 1000s projects: 1915
* Total number of results reported by MRVA after on 1000s projects: 1545
* Disregarding the barrier I added, we lose 58 (source, sink)-pairs. They all seem FPs related to pointer/pointee confusion in DTT.
* Not disregarding the barrier, we gain 82 (source, sink)-pair. I spot checked these, and they look reasonable. I'm not too worried about this number anyway, since this is a medium precision query.

Still running a MRVA experiment to see how many (source, sink)-pairs we lose when we do not disregard the barrier.

### CODE REVIEW

**Review by MathiasVP — APPROVED** [review]
LGTM if you're happy with the MRVA results. I think the query as it's written now makes a lot of sense 😍.

File: cpp/ql/src/Security/CWE/CWE-114/UncontrolledProcessOperation.ql [hunk 1]
```diff
@@ -14,25 +14,43 @@
...
+import semmle.code.cpp.ir.IR
+import Flow::PathGraph
 
-predicate isProcessOperationExplanation(Expr arg, string processOperation) {
+predicate isProcessOperationExplanation(DataFlow::Node arg, string processOperation) {
   exists(int processOperationArg, FunctionCall call |
     isProcessOperationArgument(processOperation, processOperationArg) and
     call.getTarget().getName() = processOperation and
-    call.getArgument(processOperationArg) = arg
+    call.getArgument(processOperationArg) = [arg.asExpr(), arg.asIndirectExpr()]
```
[the remark this answers is not in the document]
Comment (jketema) [reply cpp/ql/src/Security/CWE/CWE-114/UncontrolledProcessOperation.ql:26]: Note that this predicate is used on the sink-side, not the source-side. What you're saying does apply to the `not node instanceof DataFlow::ExprNode` we have in the source predicate below.

[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-114/UncontrolledProcessOperation.ql:26]: Oops, sorry. Yes, I meant to comment on the source-side. I don't know why I put the comment on this line of code 😂

File: cpp/ql/src/Security/CWE/CWE-114/UncontrolledProcessOperation.ql [hunk 2]
```diff
@@ -14,25 +14,43 @@
...
-  isProcessOperationExplanation(arg, processOperation) and
-  taintedWithPath(source, arg, sourceNode, sinkNode)
-select arg, sourceNode, sinkNode,
+  source = sourceNode.getNode() and
+  sink = sinkNode.getNode() and
+  isProcessOperationExplanation(sink, processOperation) and
+  Flow::flowPath(sourceNode, sinkNode)
+select sink, sourceNode, sinkNode,
   "The value of this argument may come from $@ and is being passed to " + processOperation + ".",
   source, source.toString()
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-114/UncontrolledProcessOperation.ql:56]: Indeed. For example here: https://github.com/github/codeql/pull/14784/files#diff-6659098746be150a26dcb4f6677d331cbea3ac76fa9945220113d2f6c624adc8R120

[the remark this answers is not in the document]
Comment (jketema) [reply cpp/ql/src/Security/CWE/CWE-114/UncontrolledProcessOperation.ql]: Done. I think we source types might need some further tuning, but that can be done later.

[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-114/UncontrolledProcessOperation.ql]: Agreed. But this in it self LGTM!

**Review by MathiasVP — APPROVED** [review]

**Comment by jketema** [comment]
> * Disregarding the barrier I added, we lose 58 (source, sink)-pairs. They all seem FPs related to pointer/pointee confusion in DTT.

Not disregarding the barrier it's 165 (source, sink) pairs

**Comment by jketema** [comment]
Rebased for internal PR purposes.

**Review by MathiasVP — APPROVED** [review]

**Pull request merged**